### PR TITLE
Auto OpSchema for trace_only functions | feat(op_schema)

### LINKED ISSUE #630: Generate opschemas for traced only functions
The relevant code is below. I wonder if there is a way to do it without `eval`

https://github.com/microsoft/onnx-script/blob/a9fc6ca065aa03d905b9abf603a8eedefb023eed/onnxscript/converter.py#L1299-L1353

### PR DESCRIPTION
**Pull request by justinchuby**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #698
* __->__ #674

This PR implements auto OpSchema generation for trace_only function as well. It leverages onnxscript's converter to create the same function IR as we do in `OnnxFunction`, but without translating the body.

We created an `TracedOnnxFunction` class to expose a similar interface as `OnnxFunction`, and an `OpLike` protocol to standardize them.

- Creates the `OpLike` protocol that defines common attributes and methods for `Op`, `OnnxFunction` and `TracedOnnxFunction` so we can assume a common interface.
- Implement `param_schemas` for `TracedOnnxFunction`
- Refactor `param_schemas` to extract common logic.
- Removes `is_single_op` from `Op` because it is unused.
- Refactor ast logic from `main.py` to `onnxscript/_internal/ast_utils.py`
- The change is tested on all the existing trace_only functions.

Fixes #630

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/tracing.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        converter = ons_converter.Converter(
+            opset=self._opset,
+            global_names=global_names,
+            source=src,
+        )
+
+        function_ir = converter.translate_function_signature(func_ast)
+
+        # Find all distinct types in the inputs and outputs
+        distinct_types = {arg.typeinfo for arg in function_ir.inputs}.union(
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/tracing.py:97]: Duplicated code starting here. May be able to extract into `op_schema_from_function_ir`

**Comment by justinchuby** [comment]
Need to fix: param order

File: onnxscript/function_libs/torch_lib/tracing.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/tracing.py]: Done

File: onnxscript/function_libs/torch_lib/registration.py [hunk 2]
```diff
@@ -81,12 +82,14 @@ def torch_op(
...
+        # Compile the function
+        custom_opset = onnxscript.values.Opset(domain="onnxscript.atenlib", version=1)
+
+        processed_func: onnxscript.OnnxFunction | tracing.TraceOnlyFunction
         if trace_only:
-            processed_func = func
+            processed_func = tracing.TraceOnlyFunction(custom_opset, func)
         else:
-            # Compile the function
-            custom_opset = onnxscript.values.Opset(domain="onnxscript.atenlib", version=1)
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/registration.py:89]: It was move to outside the if

[the remark this answers is not in the document]
Comment (titaiwangms) [reply onnxscript/function_libs/torch_lib/registration.py:89]: Oops

File: onnxscript/function_libs/torch_lib/tracing.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    src = textwrap.dedent(src)
+    top_level_ast = ast.parse(src)
+    assert isinstance(top_level_ast, ast.Module)
+    assert len(top_level_ast.body) == 1
+    f_ast = top_level_ast.body[0]
+    assert isinstance(f_ast, ast.FunctionDef)
+    return src, f_ast
+
+
+class TraceOnlyFunction:
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/tracing.py:36]: The notion of a trace only function is a torchlib notion. I am not sure if or at which stage do we want to make it an onnxscript thing.

File: onnxscript/values.py [hunk 4]
```diff
@@ -5,11 +5,13 @@
...
 
 import dataclasses
+import inspect
 import logging
 import types
+import typing
 from enum import IntFlag
 from typing import _GenericAlias  # type: ignore[attr-defined]
-from typing import Any, Optional, Sequence
+from typing import Any, Optional, Protocol, Sequence
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/values.py:15]: Good catch. Let me move it to the next pr

File: onnxscript/function_libs/torch_lib/tracing.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnxscript/function_libs/torch_lib/tracing.py:36]: I think having trace-only functions in onnxscript makes sense. Not suggesting that it be done in this PR, given that you already have a fairly complex stack of PRs, but just clarifying that it is a good idea at some point. Specifically, I think it ends up providing a spox-like functionality, which would be good to have.

[the remark this answers is not in the document]
Comment (titaiwangms) [reply onnxscript/function_libs/torch_lib/tracing.py:36]: And also this guy will be an OpLike, which will be defined in values.py. Without having this guy in values.py, OpLike seems redundant IMO.

[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/tracing.py:36]: Nice. I would like to reason about this and move it in a separate PR after this stack if that sounds good to both of you. Some preliminary thoughts on the hierarchy: Op -> ComposedOp -> {OnnxFunction (Scriptable), TraceOnlyFunction}

[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/tracing.py:36]: https://github.com/microsoft/onnxscript/issues/694

File: onnxscript/function_libs/torch_lib/tracing.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        closure = inspect.getclosurevars(self._func)
+        global_names = module.__dict__.copy()
+        global_names.update(closure.nonlocals)
+        converter = ons_converter.Converter(
+            opset=self._opset,
+            global_names=global_names,
+            source=src,
+        )
+
+        function_ir = converter.translate_function_signature(func_ast)
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_lib/tracing.py:93]: It looks scary at first, but I figure this is just parsing out the signature part from the ast?

And `op_schema_from_function_ir` below just checks the signature part of the function_ir? So it doesn't need nor look at stmts?

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/tracing.py:93]: That's correct

**Review by titaiwangms — APPROVED** [review]

File: onnxscript/values.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/values.py:15]: Combined

File: onnxscript/values.py [hunk 6]
```diff
@@ -370,11 +456,6 @@ def __init__(
         self._param_schemas: Optional[tuple[ParamSchema, ...]] = None
         self._opschema: Optional[onnx.defs.OpSchema] = None
 
-    @property
-    def name(self):
```
Comment (justinchuby) [inline onnxscript/values.py:374]: Defined by parent

**Review by gramalingam — APPROVED** [review]

File: onnxscript/function_libs/torch_lib/tracing.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/tracing.py]: Created the `TracedOnnxFunction` class

**Pull request merged**